lib/addressable/idna/native2.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+
+#--
+# Copyright (C) Bob Aman
+#
+#    Licensed under the Apache License, Version 2.0 (the "License");
+#    you may not use this file except in compliance with the License.
+#    You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#++
+
+# libidn2 implementing IDNA2008+TR46
+require "ffi"
+
+module Addressable
+  module IDNA
+    extend FFI::Library
+
+    ffi_lib ["idn2", "libidn2.0", "libidn2.so.0"]
+    # ffi_lib "idn3"
+
+    attach_function :idn2_to_ascii_8z, %i[string pointer int], :int
+    attach_function :idn2_to_unicode_8z8z, %i[string pointer int], :int
+    attach_function :idn2_strerror, [:int], :string
+    attach_function :idn2_free, [:pointer], :void
+
+    IDN2_TRANSITIONAL = 4
+    IDN2_NONTRANSITIONAL = 8
+
+    def self.to_ascii(value)
+      return value if value.ascii_only?
+      pointer = FFI::MemoryPointer.new(:pointer)
+      res = idn2_to_ascii_8z(value, pointer, IDN2_NONTRANSITIONAL)
+      # Fallback to Transitional mode in case of disallowed character
+      res = idn2_to_ascii_8z(value, pointer, IDN2_TRANSITIONAL) if res != 0
+      raise "libidn2 failed to convert \"#{value}\" to ascii (#{idn2_strerror(res)})" if res != 0
+      result = pointer.read_pointer.read_string
+      idn2_free(pointer.read_pointer)
+      result
+    end
+
+    def self.to_unicode(value)
+      pointer = FFI::MemoryPointer.new(:pointer)
+      res = idn2_to_unicode_8z8z(value, pointer, IDN2_NONTRANSITIONAL)
+      return value if res != 0
```

Comment:
Something to change in a major version bump?

Reply:
This one and 2 others would be failing if I raise an error here:
```
  1) Addressable::IDNA when using the libidn2 native implementation (ffi) it should behave like converting from ASCII to unicode should convert 'AcinusFallumTrompetumNullunCreditumVisumEstAtCuadLongumEtCefallum.com' correctly
     Failure/Error: raise "libidn2 failed to convert \"#{value}\" to unicode (#{idn2_strerror(res)})" if res != 0
     
     RuntimeError:
       libidn2 failed to convert "AcinusFallumTrompetumNullunCreditumVisumEstAtCuadLongumEtCefallum.com" to unicode (domain label longer than 63 characters)
     Shared Example Group: "converting from ASCII to unicode" called from ./spec/addressable/idna_spec.rb:321
     # ./lib/addressable/idna/libidn2.rb:52:in `to_unicode'
     # ./lib/addressable/idna.rb:30:in `to_unicode'
     # ./spec/addressable/idna_spec.rb:160:in `block (2 levels) in <top (required)>'

  2) Addressable::IDNA when using the libidn2 native implementation (ffi) it should behave like converting from ASCII to unicode should return the identity conversion when punycode decode fails
     Failure/Error: raise "libidn2 failed to convert \"#{value}\" to unicode (#{idn2_strerror(res)})" if res != 0
     
     RuntimeError:
       libidn2 failed to convert "xn--zckp1cyg1.sblo.jp" to unicode (string contains invalid punycode data)
     Shared Example Group: "converting from ASCII to unicode" called from ./spec/addressable/idna_spec.rb:321
     # ./lib/addressable/idna/libidn2.rb:52:in `to_unicode'
     # ./lib/addressable/idna.rb:30:in `to_unicode'
     # ./spec/addressable/idna_spec.rb:164:in `block (2 levels) in <top (required)>'

  3) Addressable::IDNA when using the libidn2 native implementation (ffi) it should behave like converting from ASCII to unicode should return the identity conversion when the ACE prefix has no suffix
     Failure/Error: raise "libidn2 failed to convert \"#{value}\" to unicode (#{idn2_strerror(res)})" if res != 0
     
     RuntimeError:
       libidn2 failed to convert "xn--...-" to unicode (string contains invalid punycode data)
     Shared Example Group: "converting from ASCII to unicode" called from ./spec/addressable/idna_spec.rb:321
     # ./lib/addressable/idna/libidn2.rb:52:in `to_unicode'
     # ./lib/addressable/idna.rb:30:in `to_unicode'
     # ./spec/addressable/idna_spec.rb:169:in `block (2 levels) in <top (required)>'
```
The last two are invalid punycode and the first one is invalid DNS length (https://datatracker.ietf.org/doc/html/rfc1034#section-3.1).

`libidn1` also raise in the first case but we have this workaround to explicitely allow for > 63 bytes labels:
```ruby
    def self.to_ascii(value)
      value.to_s.split('.', -1).map do |segment|
        if segment.size > 0 && segment.size < 64
          IDN::Idna.toASCII(segment, IDN::Idna::ALLOW_UNASSIGNED)
        elsif segment.size >= 64
          segment
        else
          ''
        end
      end.join('.')
    end
```
Looks like this was made in c73810f4823c4dfac1f0568073623eb7922f526d to make it more consistent with `pure`. Didn't see any issue attached though. 

So I suppose if we make `libidn2` stricter, which means basically:
```ruby
raise "libidn2 failed to convert \"#{value}\" to unicode (#{idn2_strerror(res)})" if res != 0
```
instead of
```ruby
return value if res != 0
```
We would need to remove these workarounds and make all implementations rejects these domains in the same way.
Which does sound like the way to go IMO but of course could break some use-cases for people who need to handle such "slightly" invalid domains.

Reply:
Yeah, I'm thinking c73810f4823c4dfac1f0568073623eb7922f526d was the wrong call, it should have been the pure version that changed? 

I'm down with making too long names invalid, question is, does it need to happen in a major version? 🤔 c73810f4823c4dfac1f0568073623eb7922f526d happened in 2.5.0

* [this wrapper wants it](https://github.com/twingly/twingly-url/issues/141)
* Chrome doesn't try to visit `AcinusFallumTrompetumNullunCreditumVisumEstAtCuadLongumEtCefallum.com`, it makes a search engine search for me
* ```
  $ curl -s -v AcinusFallumTrompetumNullunCreditumVisumEstAtCuadLongumEtCefallum.com
  * Could not resolve host: AcinusFallumTrompetumNullunCreditumVisumEstAtCuadLongumEtCefallum.com
  * Closing connection 0
  ```

The other two are harder, I don't know what's correct. They come from 295d7baec1281976434e9061b8065979b3f38c83 (`The spec says `toUnicode` returns the identity conversion on bad decodes.`) and cf8cc7d0b9c8d2df9caeff90a063864c3e68d940 (`Correctly handle the case where an ACE prefix has no suffix.`) – probably not good changing these in a non-major version

What if we introduce a "strict mode"? `Addressable::IDNA.strict_mode` (the user can change it) that's `false` in the next non-major version, keeping current behaviour, and `true` in the next major version meaning we raise these errors.

Reply:
> Yeah, I'm thinking https://github.com/sporkmonger/addressable/commit/c73810f4823c4dfac1f0568073623eb7922f526d was the wrong call, it should have been the pure version that changed?

That is what I think too.

Why not for the `strict_mode` (which would basically–if falsy–hide the errors and return the original name unchanged), this could even be implemented at `Addressable::IDNA` level so it supports all backends. Usually I would say this can be done on the client side so it's client responsibility, but here we're talking about one internal component of Addressable and people may want to normalize a URL even if the hostname is invalid for example. So I agree to add an option for that.

I'll try to do that.

Reply:
Ok so after giving this a try, it's not as easy as expected of course. If I add a `strict_mode` and make things right for `libidn2` (so let the native exceptions go when the string is invalid), then we get consistency problems with the other backends of course:
1. `libidn1` (even after removing the change from c73810f4823c4dfac1f0568073623eb7922f526d) does **not** raise when the label is > 63 bytes and we're converting from unicode to ascii (only the other way around). It's not a big deal (it's less strict than libidn2) but it's inconsistent so I can't just spec this behavior on all backends and would need to spec this only for `libidn2`, or we could add this exception ourselves in ruby.
2. `libidn1` does **not** raise when the punnycode is invalid like `libidn2` (I tried other flags but none of them make it raise in these two cases). Also not a big deal IMO, it's just that the strict mode will be more strict with libidn2 than libidn1. I don't see any easy way to fix that in ruby.
3.  The current `pure` implementation (even after removing the change from cf8cc7d0b9c8d2df9caeff90a063864c3e68d940) does not raise (it returns a wrong result instead), I can make it raise with small changes. For the segment length no check is made at all so we would need to add them too here if we want to be consistent, which does not look easy (it skips conversion anyway for ascii strings so we can't just add a length check in the segment loop, etc...).

So quick question here: do you think it's ok if the `strict_mode` has a different strictness depending on the backend? or shall we drop this idea and keep only the more tolerating mode for all?

Reply:
What about this idea: as "strict mode" is something new, it could be a thing only for libidn2 

Does that make sense?

Reply:
Yep, I just added the strict mode in a5e87c6824ff927adc05db4d21103fccb4ce3277.
It applies mostly to `libidn2` at the moment, though I implemented it at global level and still removed the change in libidn1 to simplify the code and make it slightly stricter (when the option is true of course).

Let me know what you think :)

Reply:
Looks good to me :) 

Need to take the time to look at/play with this PR as a whole, not sure exactly when that will be though

Reply:
Ok thanks, I've run again my 150k URL through the new "strict-mode" conversion, and it found a couple invalid ones which I double checked and are really invalid (none of them ever worked) so that is good for me. This would actually be even better for me to be able to reject those before creation so I'll probably use `strict_mode=true`.

It raised one question for me though, at the moment I rescued `Addressable::URI::InvalidURIError` to detect invalid URLs (which is theoretically not enough as we could also have `Addressable::IDNA::PunycodeBadInput`, `Addressable::IDNA::PunycodeBigOutput`, `Addressable::IDNA::PunycodeOverflow` with the Pure implementation I'm using but I've never got those). And now there can also be `Addressable::IDNA::Error` with `libidn2`.

That is fine for me and I can start rescuing `Addressable::IDNA::Error` too (which is parent to the others Pure-specific exeptions now so good enough) but I'm thinking about other users who may not be aware of this, it could be a bit surprising. You already had this problem with https://github.com/twingly/twingly-url/issues/112 so what would be your preference? I am thinking about these options:
1. Keep it like this, and people can rescue `Addressable::IDNA::Error` for all IDNA related errors (that would simplify a bit your twingly fix by the way and make it more future-proof)
2. Make Addressable::URI rescue the IDNA exceptions, and re-raise it as `Addressable::URI::InvalidURIError`, this way the API has more limited boundaries which means only one exception to rescue for users (unless they are using the IDNA module directly of course).
3. Alternativaly I just thought about this little hack which sounds easy and nice yet does not _feels_ right :upside_down_face: : make `Addressable::IDNA::Error` inherit from `Addressable::URI::InvalidURIError` instead of `StandardError`. This way people who do `rescue Addressable::URI::InvalidURIError` are actually already covered and we don't have to rescue/re-raise or change anything else. It kinda make sense as if the IDNA host is invalid it definitely means the URL is invalid, and I can't find any counter-example where this would cause trouble.

Reply:
I think 3. makes sense, it should give users the most flexibility 

Can we make that change now in the next minor version this would go into, or could it be considered a breaking change? I'm leaning towards it not being a breaking change. WDYT?

Reply:
I just tried it in ccbcb9e419de67ed17630908f229bf03e41acecf , it makes sense to me too though I'm not a big fan of leaking some of "URI" into "IDNA" (IDNA being supposedly a dependency to URI), still feels wrong but works fine :)

I think this can be made in the same minor yes, I don't see anything breaking in here as we kept the default being `safe_mode=false` and the existing exceptions haven't changed class (just parent).

Reply:
Alternative to \3., with the same outcome but without changing the inheritance would be to use `#extend`: https://github.com/twingly/twingly-url/pull/106#issuecomment-258642114

Would still be leaking but maybe a little bit cleaner? :)

Reply:
Interesting suggestion, I do like the fact that this solution being run-time, we can use this class even if it's not defined yet. Unfortunately it doesn't work with classes, only modules: `wrong argument type Class (expected Module)` if I do `error.extend(Addressable::URI::InvalidURIError)`. In your example `Twingly::URL::Error` is actually a module. And if we need to change `Addressable::URI::InvalidURIError` to be a module this would complexify the rest :/

Looking again at `uri.rb` I see `IDNA` is used only twice (in `normalized_host` for `to_ascii` and in `display_uri` for `to_unicode`), so option 2 which is to re-wrap the error here doesn't sound too complicated either.

I just gave this option a try in 9eb3910a6376fa2c36ad0a9ab04285f49905a761 and I think I actually prefer this one. No hierarchy issue here, every module/class only deal with its own exceptions. The wrapped exception are properly identified and the `cause` attribute contains the previous exception (with specs for that just in case), this means backtrace and history is complete for bug tracker. People doing `rescue Addressable::URI::InvalidURIError` are covered the same way.

I also added by the way specs for the case of invalid IDNA hostname at URI level this time (I couldn't find any at the moment). 3 of them for the Pure implementation have been marked pending because they are returning garbage at the moment (implementation makes up unicode characters from invalid input).

And I also fixed the `libidn1` exception handling, which was still letting `IDN::Idna::IdnaError` exception up so not handled properly (I missed it earlier because there was no spec on this case), now it's raising `Addressable::IDNA::Error` like the other backends (+spec)

Sorry for the long back and forth, let me know what you think about this one ;)

Reply:
Oh sorry for that, I did know about the `module` thing but forgot before I posted, oh well

Yes, I like https://github.com/sporkmonger/addressable/commit/9eb3910a6376fa2c36ad0a9ab04285f49905a761 too :) Thanks for extending the spec coverage

Reply:
Necro-ing a little bit, but wanted to weigh in on the start of the conversation. I think it's important to offer a mechanism that's permissive in what it accepts. It's literally the reason I wrote Addressable in the first place, because the standard library doesn't take this approach to parsing and I couldn't parse URIs that were openable in a browser, leading to surprise from end users. There are often cases where failing with an exception will mean that there's no graceful way to get partial information. For instance, something might be very wrong with the encoding in the hostname, but if the library's user was only trying to retrieve the path value, the invalid URI exception is rather obstructive to that goal.

On the other hand, you're absolutely right that there are cases where the opposite is preferred and strict parsing is preferable. My view is these should simply be handled by different methods rather than changing the behavior for the whole library in a major version rev.

Reply:
@sporkmonger thanks for your feedback.

I understand and agree, the concensus that @dentarg and I reached further down this discussion was to introduce the `Addressable::IDNA.strict_mode = true` option (**default to false**) so that people can choose if they want stricter parsing but otherwise it's lax as before.

Of course if you prefer different methods instead of an option, we can probably do that instead. Though if there's different methods in IDNA modules for both behaviors, we would also need to mirror that in the URI module because the methods people call usually are here. I haven't checked the whole public API recently but I'm concerned this may create a lot of new methods. I also thought about people using gems _which depends on adressable_ : if we're using different methods, the end-user won't be able to change the behavior. I can have a deeper look if you want me to :+1: